src/backend/distributed/metadata/metadata_cache.c

```diff
@@ -996,7 +996,7 @@ LookupDistObjectCacheEntry(Oid classid, Oid objid, int32 objsubid)
 	}
 
 	systable_endscan(pgDistObjectScan);
-	relation_close(pgDistObjectRel, AccessShareLock);
+	relation_close(pgDistObjectRel, NoLock);
```

Comment:
These locks are only acquired if we have a cache miss. Even if we would acquire them for cache hits it would only protect us against concurrent ALTER/DROP EXTENSION.

I think the decision to release the lock makes sense from the perspective that it aligns the behaviour between cache hit and cache miss.

There is a general question of: What is triggering the invalidations we sometimes see during INSERT..SELECT with re-partitioning? Perhaps we should obtain shard metadata locks earlier? Do we obtain them at all?

Reply:
```
create publication asdf for all tables;
alter publication asdf set (publish = 'insert');
```
`ALTER` will invalidate all caches. The publication could also be over 5000 tables that aren't related to citus

With proper locking the content when refreshed should be the same, & invalidation should be rare, but never impossible

I created a branch that tests how we handle always refreshing cache: https://github.com/citusdata/citus/commit/6c5c94b85012b4a0448d92c8fedb901efe83408a

Lots of errors that look like the weird shard id errors being looked for. I also found a case by inspection:

https://github.com/citusdata/citus/blob/d80baa35574ed8d010465bcfc1e1514c43f54494/src/backend/distributed/planner/insert_select_planner.c#L256

https://github.com/citusdata/citus/blob/d80baa35574ed8d010465bcfc1e1514c43f54494/src/backend/distributed/planner/insert_select_planner.c#L431

We pass a shardInterval from cache into `RouterModifyTaskForShardInterval` which if the the cache entry is invalid will cause the shardInterval we passed in to become an invalid pointer